src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/TargetSyntaxMessageLoader.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers
+{
+    public static class TargetSyntaxMessageLoader
+    {
+        /// <summary>
+        /// The suffix that api/message map files are expected to use.
+        /// </summary>
+        private const string TargetSyntaxMessageFileSuffix = ".apitargets";
+        private static readonly char[] NewLineCharacters = new[] { '\n', '\r' };
+        private static readonly char[] IdMessageDelimiters = new[] { ':' };
+        private static readonly char[] ApiSyntaxDelimiters = new[] { ',' };
+
+        /// <summary>
+        /// Load target syntax messages from additional files.
+        /// </summary>
+        /// <param name="additionalTexts">The additional texts to parse for type mappings.</param>
+        /// <returns>Target syntax messages as defined in *.apitargets files in the project's additional files.</returns>
+        public static IEnumerable<TargetSyntaxMessage> LoadMappings(ImmutableArray<AdditionalText> additionalTexts)
+        {
+            foreach (var file in additionalTexts)
+            {
+                if (file.Path.EndsWith(TargetSyntaxMessageFileSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    var messageMaps = LoadMappings(file.GetText()?.ToString() ?? string.Empty);
+                    if (messageMaps is null)
+                    {
+                        continue;
+                    }
+
+                    foreach (var messageMap in messageMaps)
+                    {
+                        yield return messageMap;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Load target syntax messages from a serialized string.
+        /// </summary>
+        /// <param name="serializedContents">A collection of target syntax messages stored as strings.</param>
+        /// <returns>An enumerable of TargetSyntaxMessages corresponding to those serialized in the serializedContents argument.</returns>
+        public static IEnumerable<TargetSyntaxMessage>? LoadMappings(string serializedContents)
```

Comment:
Can this just return an empty enumerable if it fails?

Reply:
It could, but I feel like throwing might be preferable here because otherwise the analyzer might silently just not do anything in the case of a malformed file. Throwing will make the failure more visible but, since it's in an analyzer, I think it still won't block anything.

Reply:
In that case, there should be an analyze to detect malformed files.

Reply:
I'll have it return an empty enumerable for now and we can look into adding an analyzer for malformed apitargets files in the future if it seems like it would help.